### PR TITLE
Batch-fetch GitLab Project visibility when computing permissions

### PR DESCRIPTION
**Pull request by beyang**
This PR adds an optimization to the GitLab OAuth authz implementation, wherein permissions for GitLab projects ("project" is what GitLab calls a repository) are fetched in batches of 100 in certain cases, instead of individually. These certain cases are those in which the projects we're fetching have visibility `public` or `internal` (i.e., not `private`). This should make permissions fetching much faster for customers using GitLab OAuth authz and have predominantly `internal` or `public` repositories on Sourcegraph.

This also adds a site config option `experimentalFeatures > debug.log > extsvc.gitlab` that can be set to `true` to turn on debug tracing of GitLab API requests in a running Sourcegraph instance. This should help with future debugging of permissions performance issues.

Please see my inline comments for more explanation of what's in this PR.

**Post-merge TODOs**
- [ ] Cut patch release for customer
- [ ] Update CHANGEOG

### CODE REVIEW

File: cmd/frontend/authz/perms.go [hunk 1]
```diff
@@ -58,3 +58,18 @@ const (
 	ProviderBitbucketServer ProviderType = bitbucketserver.ServiceType
 	ProviderSourcegraph     ProviderType = "sourcegraph"
 )
+
+// RepoPermsSort sorts a slice of RepoPerms to guarantee a stable ordering.
```
Comment (beyang) [inline cmd/frontend/authz/perms.go]: Utility sorter, to ensure stable sort ordering in test results.

File: enterprise/cmd/frontend/internal/authz/gitlab/gitlab.go [hunk 2]
```diff
@@ -78,9 +78,19 @@ func newAuthzProvider(a *schema.GitLabAuthorization, instanceURL, token string,
 			return nil, fmt.Errorf("Did not find authentication provider matching %q. Check the [**site configuration**](/site-admin/configuration) to verify an entry in [`auth.providers`](https://docs.sourcegraph.com/admin/auth) exists for %s.", instanceURL, instanceURL)
 		}
 
+		minBatchThreshold := 200
```
Comment (beyang) [inline enterprise/cmd/frontend/internal/authz/gitlab/gitlab.go]: `minBatchThreshold` and `maxBatchRequests` are documented in the `MinBatchThreshold` and `MaxBatchRequests` struct field docs. They are set here from the GitLab external services config. The default values of 200 and 300 means "if we find ourselves in a situation where we need to fetch more than 200 Projects from GitLab (at the cost of 200 API requests), let's first try fetching all the Projects accessible to the given user up to 300 * 100 (at a cost of 300 API requests)".

File: enterprise/cmd/frontend/internal/authz/gitlab/gitlab_oauth.go [hunk 3]
```diff
@@ -80,11 +95,11 @@ func (p *OAuthAuthzProvider) RepoPerms(ctx context.Context, account *extsvc.Exte
 		accountID = account.AccountID
 	}
 
-	remaining := repos
-	perms := make([]authz.RepoPerms, 0, len(remaining))
+	remaining := make([]*types.Repo, 0, len(repos))
```
Comment (beyang) [inline enterprise/cmd/frontend/internal/authz/gitlab/gitlab_oauth.go]: `remaining` tracks which repositories permissions remain to be computed for.

File: enterprise/cmd/frontend/internal/authz/gitlab/gitlab_oauth.go [hunk 4]
```diff
@@ -109,20 +124,65 @@ func (p *OAuthAuthzProvider) RepoPerms(ctx context.Context, account *extsvc.Exte
...
+	//
+	// This is an optimization. If we have too many repositories (GitLab calls them "projects") to
+	// fetch from the GitLab API, we try to batch-fetch all projects whose visibility is `internal`
+	// or `public` (because we can batch-fetch 100 repositories at a time this way). This is not
+	// guaranteed to be strictly better than fetching them individually, because if we batch-fest,
+	// we must batch-fest *all* repositories, not just the ones in `repos`.
+	//
+	// We cannot determine the permissions of projects with visibility `private` this way, because a
+	// project may be visible to a GitLab user, but its contents inaccessible (which means we have
+	// to issue individual API requests to request repository contents to verify permissions).
```
Comment (beyang) [inline enterprise/cmd/frontend/internal/authz/gitlab/gitlab_oauth.go]: This is the central optimization of this PR.

File: enterprise/cmd/frontend/internal/authz/gitlab/gitlab_oauth_test.go [hunk 5]
```diff
@@ -168,30 +176,39 @@ func Test_GitLab_RepoPerms(t *testing.T) {
...
 		t.Run(test.description, func(t *testing.T) {
 			for _, c := range test.calls {
-				t.Logf("Call %q", c.description)
-
-				// Recreate the authz provider cache every time, before running twice (once uncached, once cached)
-				ctx := context.Background()
-				op := test.op
-				op.MockCache = make(mockCache)
-				authzProvider := newOAuthProvider(op)
+				for _, batchingThreshold := range [][2]int{
```
Comment (beyang) [inline enterprise/cmd/frontend/internal/authz/gitlab/gitlab_oauth_test.go]: This test has been updated to verify correctness when different batching thresholds are used.

File: enterprise/cmd/frontend/internal/authz/gitlab/gitlab_oauth_test.go [hunk 6]
```diff
@@ -309,3 +326,151 @@ func Test_GitLab_RepoPerms_cache(t *testing.T) {
 		}
 	}
 }
+
+// Test_GitLab_RepoPerms_batchVisibility tests that Project visibility is fetched in batch (for
+// performance reasons)
```
Comment (beyang) [inline enterprise/cmd/frontend/internal/authz/gitlab/gitlab_oauth_test.go]: This test checks that the optimization is actually effective.

File: internal/extsvc/gitlab/client.go [hunk 7]
```diff
@@ -15,13 +15,37 @@ import (
...
 
-var requestCounter = metrics.NewRequestMeter("gitlab", "Total number of requests sent to the GitLab API.")
+var (
+	requestCounter = metrics.NewRequestMeter("gitlab", "Total number of requests sent to the GitLab API.")
+
+	// Whether debug nlogging is turned on
+	trace bool
+)
+
+func init() {
```
Comment (beyang) [inline internal/extsvc/gitlab/client.go]: Activate logging of GitLab API requests if `experimentalFeatures.debug.log.extsvc.gitlab` is `true` in site config.

File: enterprise/cmd/frontend/internal/authz/gitlab/gitlab_oauth_test.go [hunk 8]
```diff
@@ -168,30 +176,39 @@ func Test_GitLab_RepoPerms(t *testing.T) {
...
-
-				// Recreate the authz provider cache every time, before running twice (once uncached, once cached)
-				ctx := context.Background()
-				op := test.op
-				op.MockCache = make(mockCache)
-				authzProvider := newOAuthProvider(op)
+				for _, batchingThreshold := range [][2]int{
+					{0, 0},     // should NOT trigger batch-fetching visibility
+					{1, 300},   // should trigger batch-fetching visibility
+					{200, 300}, // should NOT trigger batch-fetching visibility
```
Comment (unknwon) [inline enterprise/cmd/frontend/internal/authz/gitlab/gitlab_oauth_test.go]: Ideally we should also have another case with `{1, 2}` which makes sure the max cap work, but that requires 201 mock repos, so no strong opinion on this.

File: enterprise/cmd/frontend/internal/authz/gitlab/gitlab_oauth_test.go [hunk 9]
```diff
@@ -309,3 +326,151 @@ func Test_GitLab_RepoPerms_cache(t *testing.T) {
...
+
+		perms, err := authzProvider.RepoPerms(ctx, acct(t, 1, "gitlab", "https://gitlab.mine/", "1", "oauth-u1"), repoSlice)
+		if err != nil {
+			t.Fatal(err)
+		}
+		sort.Sort(authz.RepoPermsSort(perms))
+		sort.Sort(authz.RepoPermsSort(expPerms))
+		if !reflect.DeepEqual(perms, expPerms) {
+			t.Errorf("expected %s, but got %s", asJSON(t, expPerms), asJSON(t, perms))
+		}
```
Comment (unknwon) [inline enterprise/cmd/frontend/internal/authz/gitlab/gitlab_oauth_test.go]: nit:
```go
		if diff := cmp.Diff(expPerms, perms); diff != "" {
			t.Errorf("perms: %v", diff)
		}
```

Same for other places if you see appropriate.

File: enterprise/cmd/frontend/internal/authz/gitlab/gitlab_oauth_test.go [hunk 10]
```diff
@@ -309,3 +326,151 @@ func Test_GitLab_RepoPerms_cache(t *testing.T) {
...
+		"u2/repo1":       repo("u2/repo1", gitlab.ServiceType, "https://gitlab.mine/", "20"),
+		"internal/repo1": repo("internal/repo1", gitlab.ServiceType, "https://gitlab.mine/", "981"),
+		"public/repo1":   repo("public/repo1", gitlab.ServiceType, "https://gitlab.mine/", "991"),
+	}
+	repoSlice := make([]*types.Repo, 0, len(repos))
+	for _, r := range repos {
+		repoSlice = append(repoSlice, r)
+	}
+
+	{
```
Comment (unknwon) [inline enterprise/cmd/frontend/internal/authz/gitlab/gitlab_oauth_test.go]: Could add a line of comment to state the purpose for the block of the code?

File: enterprise/cmd/frontend/internal/authz/gitlab/gitlab_oauth_test.go [hunk 11]
```diff
@@ -309,3 +326,151 @@ func Test_GitLab_RepoPerms_cache(t *testing.T) {
...
+		if !reflect.DeepEqual(expMadeListProjects, gitlabMock.madeListProjects) {
+			t.Errorf("expected madeListProjects calls to be %v, but got %v", expMadeListProjects, gitlabMock.madeListProjects)
+		}
+
+		gitlab.MockGetProject = nil
+		gitlab.MockListProjects = nil
+		gitlab.MockListTree = nil
+	}
+
+	{
```
Comment (unknwon) [inline enterprise/cmd/frontend/internal/authz/gitlab/gitlab_oauth_test.go]: Could add a line of comment to state the purpose for the block of the code?

File: internal/extsvc/gitlab/client.go [hunk 12]
```diff
@@ -15,13 +15,46 @@ import (
...
+	"gopkg.in/inconshreveable/log15.v2"
 )
 
-var requestCounter = metrics.NewRequestMeter("gitlab", "Total number of requests sent to the GitLab API.")
+var (
+	requestCounter = metrics.NewRequestMeter("gitlab", "Total number of requests sent to the GitLab API.")
+
+	// Whether debug nlogging is turned on
+	traceEnabled   bool
+	traceEnabledMu sync.RWMutex
```
Comment (unknwon) [inline internal/extsvc/gitlab/client.go]: Maybe simpler to avoid use of mutex:

```go
	traceEnabled atomic.Value
```

**Pull request merged**